firedrake/assemble.py

```diff
@@ -469,8 +470,14 @@ def base_form_assembly_visitor(self, expr, tensor, *args):
                 return sum(weight * arg for weight, arg in zip(expr.weights(), args))
             elif all(isinstance(op, firedrake.Cofunction) for op in args):
                 V, = set(a.function_space() for a in args)
-                res = sum([w*op.dat for (op, w) in zip(args, expr.weights())])
-                return firedrake.Cofunction(V, res)
+                result = firedrake.Cofunction(V)
+                for op, w in zip(args, expr.weights()):
+                    for dat_result, dat_op in zip(result.dat.split, op.dat.split):
+                        np.add(
+                            dat_result.data_ro_with_halos,
+                            w * dat_op.data_ro_with_halos,
+                            out=dat_result.data_wo_with_halos)
```

Comment:
I am not sure if make this operation `with_halos` is the right think to do.

Reply:
No. `result` is a `Cofunction`, and `Cofunction` assigning an expression will reach this code again, which leads to a maximum recursion.

Reply:
This is the right implementation but we should move it to a method `Dat.maxpy(weights, dats)`